types/if.d.ts

```diff
@@ -0,0 +1,13 @@
+import { Predicate, MergedUpdate, TruePredicate, FalsePredicate } from './types';
+export declare function uIf<TU, O>(predicate: FalsePredicate<O>, trueUpdates: TU, object: O): O;
+export declare function uIf<TU, O>(predicate: TruePredicate<O>, trueUpdates: TU, object: O): MergedUpdate<TU, O>;
+interface CurriedIf {
+    <TU, O>(predicate: TruePredicate<O>, trueUpdates: TU, object: O): MergedUpdate<TU, O>;
```

Comment:
yeah, that's an instance where I did the `(a,b) => (c)` currying, but not the `(a) => (b,c)` or `(a) => (b) => (c)` ones.

Reply:
it's a bit of an explosion... is it worth doing the true/false/boolean predicates all separately? I suppose it makes it more powerful. I wonder if you could do the currying once in a parameterized type (with the predicate type and return type as parameters) and then use that 3 times.

Reply:
Also, it may be possible to Curry a function in typeland: https://gist.github.com/donnut/fd56232da58d25ceecf1

Reply:
[This seems to work pretty well](https://www.typescriptlang.org/play/index.html#src=const%20x%20%3D%20(a%3A%20number%2C%20b%3A%20string)%20%3D%3E%20%22raspberry%22%0D%0A%0D%0Atype%20Curry2%3CF%20extends%20(a1%3A%20any%2C%20a2%3A%20any)%20%3D%3E%20any%3E%20%3D%20%0D%0A%20%20%20%20F%20extends%20(a1%3A%20infer%20A1%2C%20a2%3A%20infer%20A2)%20%3D%3E%20infer%20R%20%3F%0D%0A%20%20%20%20%7B%0D%0A%20%20%20%20%20%20%20%20(a1%3A%20A1%2C%20a2%3A%20A2)%3A%20R%0D%0A%20%20%20%20%20%20%20%20(a1%3A%20A1)%3A%20(a2%3A%20A2)%20%3D%3E%20R%0D%0A%20%20%20%20%7D%0D%0A%20%20%3A%20never%0D%0A%0D%0Adeclare%20const%20z%3A%20Curry2%3Ctypeof%20x%3E%0D%0A%0D%0Az(3)(%22hi%22)%0D%0Az(3%2C%20%22hi%22))

Reply:
[And Curry3](https://www.typescriptlang.org/play/index.html#src=const%20x%20%3D%20(a%3A%20number%2C%20b%3A%20string)%20%3D%3E%20%22raspberry%22%0D%0A%0D%0Atype%20Curry2%3CF%20extends%20(a1%3A%20any%2C%20a2%3A%20any)%20%3D%3E%20any%3E%20%3D%20%0D%0A%20%20%20%20F%20extends%20(a1%3A%20infer%20A1%2C%20a2%3A%20infer%20A2)%20%3D%3E%20infer%20R%20%3F%0D%0A%20%20%20%20%7B%0D%0A%20%20%20%20%20%20%20%20(a1%3A%20A1%2C%20a2%3A%20A2)%3A%20R%0D%0A%20%20%20%20%20%20%20%20(a1%3A%20A1)%3A%20(a2%3A%20A2)%20%3D%3E%20R%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20%3A%20never%0D%0A%20%20%0D%0Atype%20Curry3%3CF%20extends%20(a1%3A%20any%2C%20a2%3A%20any%2C%20a3%3A%20any)%20%3D%3E%20any%3E%20%3D%20%0D%0A%20%20%20%20F%20extends%20(a1%3A%20infer%20A1%2C%20a2%3A%20infer%20A2%2C%20a3%3A%20infer%20A3)%20%3D%3E%20infer%20R%20%3F%0D%0A%20%20%20%20%7B%0D%0A%20%20%20%20%20%20%20%20(a1%3A%20A1%2C%20a2%3A%20A2%2C%20a3%3A%20A3)%3A%20R%0D%0A%20%20%20%20%20%20%20%20(a1%3A%20A1%2C%20a2%3A%20A2)%3A%20(a3%3A%20A3)%20%3D%3E%20R%0D%0A%20%20%20%20%20%20%20%20(a1%3A%20A1)%3A%20Curry2%3C(a2%3A%20A2%2C%20a3%3A%20A3)%20%3D%3E%20R%3E%0D%0A%20%20%20%20%7D%0D%0A%20%20%3A%20never%0D%0A%0D%0Adeclare%20const%20z2%3A%20Curry2%3Ctypeof%20x%3E%0D%0A%0D%0Az2(3)(%22hi%22)%0D%0Az2(3%2C%20%22hi%22)%0D%0A%0D%0Aconst%20x3%20%3D%20(a%3A%20number%2C%20b%3A%20string%2C%20c%3A%20number%5B%5D)%20%3D%3E%204%0D%0Adeclare%20const%20z3%3A%20Curry3%3Ctypeof%20x3%3E%0D%0A%0D%0Az3(3%2C%20%22hi%22%2C%20%5B0%5D)%0D%0Az3(3%2C%20%22hi%22)(%5B0%5D)%0D%0Az3(3)(%22hi%22%2C%20%5B0%5D)%0D%0Az3(3)(%22hi%22)(%5B0%5D)%0D%0A), though the last comment in the gist link claims to solve it variadically, though I haven't tested it. 

I may prefer the more explicit versions, especially since updeep has its own curry2/3/4 functions already, it matches those.

Reply:
https://medium.com/@hernanrajchert/creating-typings-for-curry-using-ts-3-x-956da2780bbf is also a good reference for generic curried functions.

Reply:
Ah yes, I played with it this morning, and re-discovered the snags I got in the first place.

So

```
type Curry2<F extends (a: any, b: any) => any> =
    F extends (a: infer A, b: infer B) => infer R ?
    {
        (a: A, b: B): R
        (a: A): (b: B) => R
    }
  : never;
```

works, but it also lose information along the way.  Best thing is to go through an example. Let's take updeep's map function, and let's have

```
type Curry2<F extends (a: any, b: any) => any> =
    F extends (a: infer A, b: infer B) => infer R ?
    {
        (a: A, b: B): R
        (a: A): (b: B) => R
    }
  : never;

interface CurriedMap {
    <I,O extends object>(iteratee: I, object: O): Mapped<I,O>;
    <I,O extends object>(iteratee: I): (object: O)=> Mapped<I,O>;
};

const curried1 = wrapped as CurriedMap;
const curried2 = wrapped as Curry2<typeof map>;

let x = curried1({ foo:1 }, { foo: 2 });
x.foo; // x is of type Mapped<{foo: number},{foo: number}>, and this line is fine

let y = curried2({ foo:1 }, { foo: 2 });
y.foo;  // y is of type object, and this line reports an error because `foo` is not a know key of {}.

```

Mind you, I'm still new at Typescript and I'm not yet fully cognisant of the extents and limitations of its inference powers, but what I think happens is that  Curry2 resolves `Mapped<I,O>` in the more generic context (so can only assume that both I and O are objects, but nothing more than that), and CurriedMap keeps the return statement as `Mapped<I,O>`, which only gets resolved when needed later on (and when we have more specific information on I and O). 

I hope that I'm making some sense here. But TL;DR: having generic curried meta-types work, but they'll degrade the precision of the initial type.

Btw, one of my experiments was to totally replace the local `curry` function with lodash's, in the hope that the curried types would magically be dealt with, but I observed the same thing: the curried types would be okay, but more generic than the original types.

Reply:
Ah, that's too bad. Yeah, I'm not sure of another way to do it either. It does seem like an issue that the type parameters can't be properly inferred. I guess doing it manually is the way to go 😖

Reply:
Yeah. Fortunately up to 4 curried levels isn't that horrid to deal with, but it's still meh.

Something that makes me slightly sad is that it seems there is no good way for typescript projects to use the typescript version of their dependencies directly. For dependencies with basic types, that's okay as the `.d.ts` will match the signatures anyway, but for more complex signatures where inference would shine, that's sub-optimal. :-P